src/OneSignal.ts

```diff
@@ -92,32 +92,48 @@ export default class OneSignal {
    * @PublicApi
    */
   static async setEmail(email: string, options?: SetEmailOptions): Promise<string> {
-    if (!email)
+    if (!email) {
       throw new InvalidArgumentError('email', InvalidArgumentReason.Empty);
-    if (!isValidEmail(email))
+    }
+    if (!isValidEmail(email)) {
       throw new InvalidArgumentError('email', InvalidArgumentReason.Malformed);
-    // emailAuthHash is expected to be a 64 character SHA-256 hex hash
-    if (options && options.emailAuthHash && options.emailAuthHash.length !== 64) {
-      throw new InvalidArgumentError('options.emailAuthHash', InvalidArgumentReason.Malformed);
     }
-    await awaitOneSignalInitAndSupported();
+
+    const isIdentifierAuthHashDefined = options && !!options.identifierAuthHash;
+    const isEmailAuthHashDefined      = options && !!options.emailAuthHash;
+
+    const authHash = isIdentifierAuthHashDefined ? options.identifierAuthHash :
+    (isEmailAuthHashDefined ? options.emailAuthHash : undefined);
+
+    if (!!authHash) {
+      if (isIdentifierAuthHashDefined && isEmailAuthHashDefined) {
+        Log.error("Both `emailAuthHash` and `identifierAuthHash` provided.");
+        throw new InvalidArgumentError('options', InvalidArgumentReason.Malformed);
+      }
+      // identifierAuthHash / emailAuthHash are expected to be a 64 character SHA-256 hex hash
+      const isIdentifierAuthHashMalformed = isIdentifierAuthHashDefined && options.identifierAuthHash.length !== 64;
+      const isEmailAuthHashIsMalformed    = isEmailAuthHashDefined && options.emailAuthHash.length !== 64;
+
+      if ( isIdentifierAuthHashMalformed ) {
+        throw new InvalidArgumentError('options.identifierAuthHash', InvalidArgumentReason.Malformed);
+      }
+
+      if ( isEmailAuthHashIsMalformed ) {
+        throw new InvalidArgumentError('options.emailAuthHash', InvalidArgumentReason.Malformed);
+      }
+    }
+
     logMethodCall('setEmail', email, options);
+    await awaitOneSignalInitAndSupported();
 
     const appConfig = await Database.getAppConfig();
     const { deviceId } = await Database.getSubscription();
     const existingEmailProfile = await Database.getEmailProfile();
 
-    if (appConfig.emailAuthRequired && !(options && options.emailAuthHash)) {
-      throw new InvalidArgumentError('options.emailAuthHash', InvalidArgumentReason.Empty);
-    }
-
-    const newEmailProfile = new EmailProfile(existingEmailProfile.emailId, email);
-    if (options && options.emailAuthHash) {
-      newEmailProfile.emailAuthHash = options.emailAuthHash
-    }
+    const newEmailProfile = new EmailProfile(existingEmailProfile.emailId, email, authHash);
 
     const isExistingEmailSaved = !!existingEmailProfile.emailId;
-    if (isExistingEmailSaved && appConfig.emailAuthRequired) {
```

Comment:
Motivation: removed `emailAuthRequired` from remote config. The logic wasn't correct anyway. Whether or not this config was required shouldn't impact whether the call is an update call or a create call

Reply:
👍